src/client/opamCommands.ml

```diff
@@ -951,6 +962,12 @@ let config =
     | Some `subst, (_::_ as files) ->
       OpamGlobalState.with_ `Lock_none @@ fun gt ->
       `Ok (OpamConfigCommand.subst gt (List.map OpamFilename.Base.of_string files))
+    | Some `setopt, [var;value] ->
+      OpamGlobalState.with_ `Lock_none @@ fun gt ->
```

Comment:
Not `Lock_none` no ?

Reply:
The `write` lock is taken in the end, when it is finally needed to write (checks are ok).

Reply:
Please don't use lock upgrade unless you have to (which should only happen when you normally don't need to lock but there is a specific case, like format upgrade). Take all the locks you need ASAP (always in the same order), then discard them ASAP once you don't need them. Lock upgrade can cause deadlocks.